app/validation/validator.py

```diff
@@ -491,63 +486,79 @@ def _validate_answer_ids_present_in_schema(self, when, answer_ids_with_group_id,
-        else:  # We know we have a 'answer_ids' key so can do validation against it
-            if when['condition'] in ('contains', 'not contains'):
-                errors.append(self._error_message('The condition "{}" is not valid for an answer_count based "when" clause'
-                                                  .format(when['condition'])))
-            if len(when['answer_ids']) != len(set(when['answer_ids'])):
-                errors.append(self._error_message('Duplicate answer ids found within {} clause'.format(referenced_id)))
+        condition = when['condition']
+        valid_checkbox_conditions = ('contains any', 'contains all', 'contains', 'not contains', 'set', 'not set')
+        answer_type = answer_ids_with_group_id[when['id']]['answer']['type'] if 'id' in when else None
+
+        if answer_type == 'Checkbox':
+            if condition not in valid_checkbox_conditions:
+                errors.append(self._error_message(f'The condition `{condition}` cannot be used with `Checkbox` answer type.'))
+        elif condition in ('contains any', 'contains all', 'contains', 'not contains'):
+            errors.append(
+                self._error_message(f'The condition `{condition}` can only be used with `Checkbox` answer types. Found answer type: {answer_type}'))
 
         return errors
 
-    def _validate_comparison_id_in_when_rule(self, when, answer_ids_with_group_id, referenced_id):
+    def _validate_conditions_requiring_list_match_values_are_valid_in_when_rule(self, when, answer_ids_with_group_id, referenced_id):
         """
-        Validate that all types with comparison_id match
+        Validate conditions requiring list match values define the values or comparison id property
         :return: list of dictionaries containing error messages, otherwise it returns an empty list
         """
         errors = []
-        comparison_type = answer_ids_with_group_id[when['comparison_id']]['answer']['type']
-        id_type = answer_ids_with_group_id[when['id']]['answer']['type']
+        condition, answer_id = when['condition'], when['id']
+        conditions_requiring_list_match_values = ('equals any', 'not equals any', 'contains any', 'contains all')
 
-        if comparison_type != id_type:
-            errors.append(self._error_message('The answers used as comparison_id "{}" and answer_id "{}" in the '
-                                              '"when" clause for {} have different types'
-                                              .format(when['comparison_id'], when['id'], referenced_id)))
+        if condition in conditions_requiring_list_match_values and not when.get('values', when.get('comparison_id')):
+            errors.append(self._error_message(
+                f'Either the properties `values` or `comparison_id` is missing/empty for when condition `{condition}` with answer id `{answer_id}`'))
```

Comment:
Isn't it possible to determine which of these are missing and provide a more specific error?

Reply:
This is being removed as this is caught by schema definition anyway.